Build failure: tk-8_5

### Nixpkgs version

- Unstable (26.05)

### Steps to reproduce

```
nix build "github:NixOS/nixpkgs/nixos-unstable-small#tk-8_5"
```

### Can Hydra reproduce this build failure?

Yes, Hydra can reproduce this build failure.

### Link to Hydra build job

https://hydra.nixos.org/build/318200331

### Relevant log output

```console
build flags: -j16 SHELL=/nix/store/lw117lsr8d585xs63kx5k233impyrq7q-bash-5.3p3/bin/bash
gcc -c -O2  -pipe -m64  -Wall -fPIC -I/build/tk8.5.19/unix/../unix -I/build/tk8.5.19/unix/../generic -I/build/tk8.5.19/unix/../bitmaps -I/build/tcl8.5.19/generic -I/build/tcl8.>
gcc -c -O2  -pipe -m64  -Wall -fPIC -I/build/tk8.5.19/unix/../unix -I/build/tk8.5.19/unix/../generic -I/build/tk8.5.19/unix/../bitmaps -I/build/tcl8.5.19/generic -I/build/tcl8.>
gcc -c -O2  -pipe -m64  -Wall -fPIC -I/build/tk8.5.19/unix/../unix -I/build/tk8.5.19/unix/../generic -I/build/tk8.5.19/unix/../bitmaps -I/build/tcl8.5.19/generic -I/build/tcl8.>
gcc -c -O2  -pipe -m64  -Wall -fPIC -I/build/tk8.5.19/unix/../unix -I/build/tk8.5.19/unix/../generic -I/build/tk8.5.19/unix/../bitmaps -I/build/tcl8.5.19/generic -I/build/tcl8.>
gcc -c -O2  -pipe -m64  -Wall -fPIC -I/build/tk8.5.19/unix/../unix -I/build/tk8.5.19/unix/../generic -I/build/tk8.5.19/unix/../bitmaps -I/build/tcl8.5.19/generic -I/build/tcl8.>
gcc -c -O2  -pipe -m64  -Wall -fPIC -I/build/tk8.5.19/unix/../unix -I/build/tk8.5.19/unix/../generic -I/build/tk8.5.19/unix/../bitmaps -I/build/tcl8.5.19/generic -I/build/tcl8.>
gcc -c -O2  -pipe -m64  -Wall -fPIC -I/build/tk8.5.19/unix/../unix -I/build/tk8.5.19/unix/../generic -I/build/tk8.5.19/unix/../bitmaps -I/build/tcl8.5.19/generic -I/build/tcl8.>
gcc -c -O2  -pipe -m64  -Wall -fPIC -I/build/tk8.5.19/unix/../unix -I/build/tk8.5.19/unix/../generic -I/build/tk8.5.19/unix/../bitmaps -I/build/tcl8.5.19/generic -I/build/tcl8.>
gcc -c -O2  -pipe -m64  -Wall -fPIC -I/build/tk8.5.19/unix/../unix -I/build/tk8.5.19/unix/../generic -I/build/tk8.5.19/unix/../bitmaps -I/build/tcl8.5.19/generic -I/build/tcl8.>
gcc -c -O2  -pipe -m64  -Wall -fPIC -I/build/tk8.5.19/unix/../unix -I/build/tk8.5.19/unix/../generic -I/build/tk8.5.19/unix/../bitmaps -I/build/tcl8.5.19/generic -I/build/tcl8.>
gcc -c -O2  -pipe -m64  -Wall -fPIC -I/build/tk8.5.19/unix/../unix -I/build/tk8.5.19/unix/../generic -I/build/tk8.5.19/unix/../bitmaps -I/build/tcl8.5.19/generic -I/build/tcl8.>
gcc -c -O2  -pipe -m64  -Wall -fPIC -I/build/tk8.5.19/unix/../unix -I/build/tk8.5.19/unix/../generic -I/build/tk8.5.19/unix/../bitmaps -I/build/tcl8.5.19/generic -I/build/tcl8.>
gcc -c -O2  -pipe -m64  -Wall -fPIC -I/build/tk8.5.19/unix/../unix -I/build/tk8.5.19/unix/../generic -I/build/tk8.5.19/unix/../bitmaps -I/build/tcl8.5.19/generic -I/build/tcl8.>
gcc -c -O2  -pipe -m64  -Wall -fPIC -I/build/tk8.5.19/unix/../unix -I/build/tk8.5.19/unix/../generic -I/build/tk8.5.19/unix/../bitmaps -I/build/tcl8.5.19/generic -I/build/tcl8.>
gcc -c -O2  -pipe -m64  -Wall -fPIC -I/build/tk8.5.19/unix/../unix -I/build/tk8.5.19/unix/../generic -I/build/tk8.5.19/unix/../bitmaps -I/build/tcl8.5.19/generic -I/build/tcl8.>
gcc -c -O2  -pipe -m64  -Wall -fPIC -I/build/tk8.5.19/unix/../unix -I/build/tk8.5.19/unix/../generic -I/build/tk8.5.19/unix/../bitmaps -I/build/tcl8.5.19/generic -I/build/tcl8.>
gcc -c -O2  -pipe -m64  -Wall -fPIC -I/build/tk8.5.19/unix/../unix -I/build/tk8.5.19/unix/../generic -I/build/tk8.5.19/unix/../bitmaps -I/build/tcl8.5.19/generic -I/build/tcl8.>
gcc -c -O2  -pipe -m64  -Wall -fPIC -I/build/tk8.5.19/unix/../unix -I/build/tk8.5.19/unix/../generic -I/build/tk8.5.19/unix/../bitmaps -I/build/tcl8.5.19/generic -I/build/tcl8.>
/build/tk8.5.19/unix/../generic/tkFocus.c: In function 'TkFocusSplit':
/build/tk8.5.19/unix/../generic/tkFocus.c:1076:1: warning: old-style function definition [-Wold-style-definition]
 1076 | TkFocusSplit(winPtr)
      | ^~~~~~~~~~~~
/build/tk8.5.19/unix/../generic/tkFocus.c: In function 'TkFocusJoin':
/build/tk8.5.19/unix/../generic/tkFocus.c:1156:1: warning: old-style function definition [-Wold-style-definition]
 1156 | TkFocusJoin(winPtr)
      | ^~~~~~~~~~~
gcc -c -O2  -pipe -m64  -Wall -fPIC -I/build/tk8.5.19/unix/../unix -I/build/tk8.5.19/unix/../generic -I/build/tk8.5.19/unix/../bitmaps -I/build/tcl8.5.19/generic -I/build/tcl8.>
gcc -c -O2  -pipe -m64  -Wall -fPIC -I/build/tk8.5.19/unix/../unix -I/build/tk8.5.19/unix/../generic -I/build/tk8.5.19/unix/../bitmaps -I/build/tcl8.5.19/generic -I/build/tcl8.>
/build/tk8.5.19/unix/../generic/tkClipboard.c: In function 'ClipboardAppHandler':
/build/tk8.5.19/unix/../generic/tkClipboard.c:150:5: warning: '__builtin_strncpy' specified bound depends on the length of the source argument [-Wstringop-truncation]
  150 |     strncpy(buffer, p, length);
      |     ^
/build/tk8.5.19/unix/../generic/tkClipboard.c:142:14: note: length computed here
  142 |     length = strlen(p);
      |              ^~~~~~~~~
gcc -c -O2  -pipe -m64  -Wall -fPIC -I/build/tk8.5.19/unix/../unix -I/build/tk8.5.19/unix/../generic -I/build/tk8.5.19/unix/../bitmaps -I/build/tcl8.5.19/generic -I/build/tcl8.>
gcc -c -O2  -pipe -m64  -Wall -fPIC -I/build/tk8.5.19/unix/../unix -I/build/tk8.5.19/unix/../generic -I/build/tk8.5.19/unix/../bitmaps -I/build/tcl8.5.19/generic -I/build/tcl8.>
gcc -c -O2  -pipe -m64  -Wall -fPIC -I/build/tk8.5.19/unix/../unix -I/build/tk8.5.19/unix/../generic -I/build/tk8.5.19/unix/../bitmaps -I/build/tcl8.5.19/generic -I/build/tcl8.>
gcc -c -O2  -pipe -m64  -Wall -fPIC -I/build/tk8.5.19/unix/../unix -I/build/tk8.5.19/unix/../generic -I/build/tk8.5.19/unix/../bitmaps -I/build/tcl8.5.19/generic -I/build/tcl8.>
gcc -c -O2  -pipe -m64  -Wall -fPIC -I/build/tk8.5.19/unix/../unix -I/build/tk8.5.19/unix/../generic -I/build/tk8.5.19/unix/../bitmaps -I/build/tcl8.5.19/generic -I/build/tcl8.>
gcc -c -O2  -pipe -m64  -Wall -fPIC -I/build/tk8.5.19/unix/../unix -I/build/tk8.5.19/unix/../generic -I/build/tk8.5.19/unix/../bitmaps -I/build/tcl8.5.19/generic -I/build/tcl8.>
gcc -c -O2  -pipe -m64  -Wall -fPIC -I/build/tk8.5.19/unix/../unix -I/build/tk8.5.19/unix/../generic -I/build/tk8.5.19/unix/../bitmaps -I/build/tcl8.5.19/generic -I/build/tcl8.>
gcc -c -O2  -pipe -m64  -Wall -fPIC -I/build/tk8.5.19/unix/../unix -I/build/tk8.5.19/unix/../generic -I/build/tk8.5.19/unix/../bitmaps -I/build/tcl8.5.19/generic -I/build/tcl8.>
gcc -c -O2  -pipe -m64  -Wall -fPIC -I/build/tk8.5.19/unix/../unix -I/build/tk8.5.19/unix/../generic -I/build/tk8.5.19/unix/../bitmaps -I/build/tcl8.5.19/generic -I/build/tcl8.>
gcc -c -O2  -pipe -m64  -Wall -fPIC -I/build/tk8.5.19/unix/../unix -I/build/tk8.5.19/unix/../generic -I/build/tk8.5.19/unix/../bitmaps -I/build/tcl8.5.19/generic -I/build/tcl8.>
gcc -c -O2  -pipe -m64  -Wall -fPIC -I/build/tk8.5.19/unix/../unix -I/build/tk8.5.19/unix/../generic -I/build/tk8.5.19/unix/../bitmaps -I/build/tcl8.5.19/generic -I/build/tcl8.>
gcc -c -O2  -pipe -m64  -Wall -fPIC -I/build/tk8.5.19/unix/../unix -I/build/tk8.5.19/unix/../generic -I/build/tk8.5.19/unix/../bitmaps -I/build/tcl8.5.19/generic -I/build/tcl8.>
gcc -c -O2  -pipe -m64  -Wall -fPIC -I/build/tk8.5.19/unix/../unix -I/build/tk8.5.19/unix/../generic -I/build/tk8.5.19/unix/../bitmaps -I/build/tcl8.5.19/generic -I/build/tcl8.>
gcc -c -O2  -pipe -m64  -Wall -fPIC -I/build/tk8.5.19/unix/../unix -I/build/tk8.5.19/unix/../generic -I/build/tk8.5.19/unix/../bitmaps -I/build/tcl8.5.19/generic -I/build/tcl8.>
gcc -c -O2  -pipe -m64  -Wall -fPIC -I/build/tk8.5.19/unix/../unix -I/build/tk8.5.19/unix/../generic -I/build/tk8.5.19/unix/../bitmaps -I/build/tcl8.5.19/generic -I/build/tcl8.>
gcc -c -O2  -pipe -m64  -Wall -fPIC -I/build/tk8.5.19/unix/../unix -I/build/tk8.5.19/unix/../generic -I/build/tk8.5.19/unix/../bitmaps -I/build/tcl8.5.19/generic -I/build/tcl8.>
gcc -c -O2  -pipe -m64  -Wall -fPIC -I/build/tk8.5.19/unix/../unix -I/build/tk8.5.19/unix/../generic -I/build/tk8.5.19/unix/../bitmaps -I/build/tcl8.5.19/generic -I/build/tcl8.>
gcc -c -O2  -pipe -m64  -Wall -fPIC -I/build/tk8.5.19/unix/../unix -I/build/tk8.5.19/unix/../generic -I/build/tk8.5.19/unix/../bitmaps -I/build/tcl8.5.19/generic -I/build/tcl8.>
gcc -c -O2  -pipe -m64  -Wall -fPIC -I/build/tk8.5.19/unix/../unix -I/build/tk8.5.19/unix/../generic -I/build/tk8.5.19/unix/../bitmaps -I/build/tcl8.5.19/generic -I/build/tcl8.>
gcc -c -O2  -pipe -m64  -Wall -fPIC -I/build/tk8.5.19/unix/../unix -I/build/tk8.5.19/unix/../generic -I/build/tk8.5.19/unix/../bitmaps -I/build/tcl8.5.19/generic -I/build/tcl8.>
/build/tk8.5.19/unix/../generic/tkEntry.c: In function 'EntryValidate':
/build/tk8.5.19/unix/../generic/tkEntry.c:3184:15: error: expected identifier or '(' before 'bool'
 3184 |     int code, bool;
      |               ^~~~
/build/tk8.5.19/unix/../generic/tkEntry.c:3206:14: error: expected expression before 'bool'
 3206 |             &bool) != TCL_OK) {
      |              ^~~~
/build/tk8.5.19/unix/../generic/tkEntry.c:3215:17: error: expected ')' before '?' token
 3215 |     return (bool ? TCL_OK : TCL_BREAK);
      |            ~    ^~
      |                 )
/build/tk8.5.19/unix/../generic/tkEntry.c:3215:39: error: expected expression before ';' token
 3215 |     return (bool ? TCL_OK : TCL_BREAK);
      |                                       ^
make: *** [Makefile:1007: tkEntry.o] Error 1
make: *** Waiting for unfinished jobs....
/build/tk8.5.19/unix/../generic/tkFrame.c: In function 'TkMapTopFrame':
/build/tk8.5.19/unix/../generic/tkFrame.c:1923:1: warning: old-style function definition [-Wold-style-definition]
 1923 | TkMapTopFrame (tkwin)
      | ^~~~~~~~~~~~~
```

### Additional context

This issue also causes build failure of `github:NixOS/nixpkgs/nixos-unstable-small#pypy3`.

### System metadata

 - system: `"x86_64-linux"`
 - host os: `Linux 6.17.12-xanmod1, NixOS, 26.05 (Yarara), 26.05.20251228.652f1b0`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Lix, like Nix) 2.95.0-pre20251227-dev_fa62109
System type: x86_64-linux
Additional system types: aarch64-linux, armv6l-linux, armv7l-linux, i686-linux, loongarch64-linux, mips-linux, mips64-linux, mips64-linuxabin32, mips64el-linux, mips64el-linuxabin32, mipsel-linux, riscv32-linux, riscv64-linux, wasm32-wasi, wasm64-wasi
Features: gc, signed-caches
System configuration file: /etc/nix/nix.conf
User configuration files: /home/mix/.config/nix/nix.conf:/home/mix/.config/kdedefaults/nix/nix.conf:/nix/store/84azgsygbh75g0p7ij2mf4f7s185a1r5-plasma-workspace-6.5.4/etc/xdg/nix/nix.conf:/nix/store/ajw102q1y1n33ild2fghww5ix8cib67p-kglobalacceld-6.5.4/etc/xdg/nix/nix.conf:/nix/store/avfxjcahrbsnyzvkcwd63ln94n8961ja-baloo-6.21.0/etc/xdg/nix/nix.conf:/home/mix/.config/kdedefaults/nix/nix.conf:/etc/xdg/nix/nix.conf:/home/mix/.nix-profile/etc/xdg/nix/nix.conf:/nix/profile/etc/xdg/nix/nix.conf:/home/mix/.local/state/nix/profile/etc/xdg/nix/nix.conf:/etc/profiles/per-user/mix/etc/xdg/nix/nix.conf:/nix/var/nix/profiles/default/etc/xdg/nix/nix.conf:/run/current-system/sw/etc/xdg/nix/nix.conf
Store directory: /nix/store
State directory: /nix/var/nix
Data directory: /nix/store/9ynpa1khjqgm55rv5azz8yqxycd1k48m-lix-2.95.0-pre20251227-dev_fa62109/share`
 - nixpkgs: `/nix/store/brz8khqrz7lac6ydd7a5l62ypqm6n4zh-source`


### Notify maintainers

:(

---

**Note for maintainers:** Please tag this issue in your pull request description. (i.e. `Resolves #ISSUE`.)


### I assert that this issue is relevant for Nixpkgs

- [x] I assert that this is a bug and not a support request.
- [x] I assert that this is not a [duplicate of an existing issue](https://github.com/NixOS/nixpkgs/issues?q=is%3Aopen+is%3Aissue+label%3A%220.kind%3A+build+failure%22). 
- [x] I assert that I have read the [NixOS Code of Conduct](https://github.com/NixOS/.github/blob/master/CODE_OF_CONDUCT.md) and agree to abide by it.

### Is this issue important to you?

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

--- a/pkgs/development/libraries/tk/generic.nix
+++ b/pkgs/development/libraries/tk/generic.nix
@@ -89,6 +89,10 @@ tcl.mkTclDerivation {
 
   inherit tcl;
 
+  env = lib.optionalAttrs (lib.versionOlder tcl.version "8.6") {
+    NIX_CFLAGS_COMPILE = "-std=gnu17";
+  };
+
   passthru = rec {
     inherit (tcl) release version;
     libPrefix = "tk${tcl.release}";

